### PR TITLE
Update llama.cpp and move core processing to native code

### PR DESCRIPTION
**Pull request by dsd**
Thanks for taking the initiative on Sherpa; I was also curious about the combination of low end devices, flutter, and open source AI, and it was nice to see that you had already been working on this.

It wasn't working on my phone (Samsung Galaxy S10) due to a llama.cpp crasher, and also memory exhaustion, but with the changes made here it works rather well with new 3B models.

### CODE REVIEW

**Comment by moh21amed** [comment]
can it run 3b on mobile with 3gb ram ?

**Comment by dsd** [comment]
> can it run 3b on mobile with 3gb ram ?

Not sure, if you want to try it then there is an apk [here](https://github.com/dsd/sherpa/releases/tag/2.2.1-dsd2)
I suspect it won't work, because the 3B files I have seen are around 2GB, and I suspect your base OS is using at least 1GB RAM...

**Comment by windmaple** [comment]
Strangely I got a bunch of errors when compiling for Android

>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>.
Launching lib/main.dart on sdk gphone64 x86 64 in debug mode...
Upgrading build.gradle

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:buildCMakeDebug[arm64-v8a]'.
> Build command failed.
  Error while executing process /usr/local/google/home/wind-test/Android/Sdk/cmake/3.18.1/bin/ninja with arguments {-C /usr/local/google/home/wind-test/Desktop/project-src/sherpa/android/app/.cxx/Debug/5a1i5p1k/arm64-v8a llama llamasherpa}
  ninja: Entering directory `/usr/local/google/home/wind-test/Desktop/project-src/sherpa/android/app/.cxx/Debug/5a1i5p1k/arm64-v8a'
  [1/5] Building C object llama.cpp/CMakeFiles/ggml.dir/ggml.c.o
  /usr/local/google/home/wind-test/Desktop/project-src/sherpa/src/llama.cpp/ggml.c:1797:5: warning: implicit conversion increases floating-point precision: 'float' to 'ggml_float' (aka 'double') [-Wdouble-promotion]
      GGML_F16_VEC_REDUCE(sumf, sum);
      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  /usr/local/google/home/wind-test/Desktop/project-src/sherpa/src/llama.cpp/ggml.c:1310:41: note: expanded from macro 'GGML_F16_VEC_REDUCE'
      #define GGML_F16_VEC_REDUCE         GGML_F32Cx4_REDUCE
                                          ^
  /usr/local/google/home/wind-test/Desktop/project-src/sherpa/src/llama.cpp/ggml.c:1300:38: note: expanded from macro 'GGML_F32Cx4_REDUCE'
      #define GGML_F32Cx4_REDUCE       GGML_F32x4_REDUCE
                                       ^
  /usr/local/google/home/wind-test/Desktop/project-src/sherpa/src/llama.cpp/ggml.c:1233:11: note: expanded from macro 'GGML_F32x4_REDUCE'
      res = GGML_F32x4_REDUCE_ONE(x[0]);         \
          ~ ^~~~~~~~~~~~~~~~~~~~~~~~~~~
  /usr/local/google/home/wind-test/Desktop/project-src/sherpa/src/llama.cpp/ggml.c:1219:27: note: expanded from macro 'GGML_F32x4_REDUCE_ONE'
       vgetq_lane_f32(x, 2) +          \
       ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~
  /usr/local/google/home/wind-test/Desktop/project-src/sherpa/src/llama.cpp/ggml.c:2354:9: warning: implicit conversion increases floating-point precision: 'float' to 'ggml_float' (aka 'double') [-Wdouble-promotion]
          GGML_F16_VEC_REDUCE(sumf[k], sum[k]);
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  /usr/local/google/home/wind-test/Desktop/project-src/sherpa/src/llama.cpp/ggml.c:1310:41: note: expanded from macro 'GGML_F16_VEC_REDUCE'
      #define GGML_F16_VEC_REDUCE         GGML_F32Cx4_REDUCE
                                          ^
  /usr/local/google/home/wind-test/Desktop/project-src/sherpa/src/llama.cpp/ggml.c:1300:38: note: expanded from macro 'GGML_F32Cx4_REDUCE'
      #define GGML_F32Cx4_REDUCE       GGML_F32x4_REDUCE
                                       ^
  /usr/local/google/home/wind-test/Desktop/project-src/sherpa/src/llama.cpp/ggml.c:1233:11: note: expanded from macro 'GGML_F32x4_REDUCE'
      res = GGML_F32x4_REDUCE_ONE(x[0]);         \
          ~ ^~~~~~~~~~~~~~~~~~~~~~~~~~~
  /usr/local/google/home/wind-test/Desktop/project-src/sherpa/src/llama.cpp/ggml.c:1219:27: note: expanded from macro 'GGML_F32x4_REDUCE_ONE'
       vgetq_lane_f32(x, 2) +          \
       ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~
  2 warnings generated.
  [2/5] Building CXX object llamasherpa/CMakeFiles/llamasherpa.dir/llamasherpa.cpp.o
  FAILED: llamasherpa/CMakeFiles/llamasherpa.dir/llamasherpa.cpp.o 
  /usr/local/google/home/wind-test/Android/Sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++ --target=aarch64-none-linux-android21 --gcc-toolchain=/usr/local/google/home/wind-test/Android/Sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64 --sysroot=/usr/local/google/home/wind-test/Android/Sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/sysroot -Dllamasherpa_EXPORTS -I/usr/local/google/home/wind-test/Desktop/project-src/sherpa/src/llamasherpa/../llama.cpp -I/usr/local/google/home/wind-test/Desktop/project-src/sherpa/src/llama.cpp/. -g -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security   -fno-limit-debug-info  -fPIC -MD -MT llamasherpa/CMakeFiles/llamasherpa.dir/llamasherpa.cpp.o -MF llamasherpa/CMakeFiles/llamasherpa.dir/llamasherpa.cpp.o.d -o llamasherpa/CMakeFiles/llamasherpa.dir/llamasherpa.cpp.o -c /usr/local/google/home/wind-test/Desktop/project-src/sherpa/src/llamasherpa/llamasherpa.cpp
  /usr/local/google/home/wind-test/Desktop/project-src/sherpa/src/llamasherpa/llamasherpa.cpp:94:8: error: unknown type name 'llama_model'; did you mean 'llama_token'?
  static llama_model *model;
         ^~~~~~~~~~~
         llama_token
  /usr/local/google/home/wind-test/Desktop/project-src/sherpa/src/llamasherpa/../llama.cpp/llama.h:80:17: note: 'llama_token' declared here
      typedef int llama_token;
                  ^
  /usr/local/google/home/wind-test/Desktop/project-src/sherpa/src/llamasherpa/llamasherpa.cpp:112:5: error: use of undeclared identifier 'llama_init_backend'
      llama_init_backend(false);
      ^
  /usr/local/google/home/wind-test/Desktop/project-src/sherpa/src/llamasherpa/llamasherpa.cpp:119:13: error: no member named 'n_batch' in 'llama_context_params'
      lparams.n_batch      = n_batch;
      ~~~~~~~ ^
  /usr/local/google/home/wind-test/Desktop/project-src/sherpa/src/llamasherpa/llamasherpa.cpp:122:14: error: use of undeclared identifier 'llama_load_model_from_file'
      model  = llama_load_model_from_file(model_path, lparams);
               ^
  /usr/local/google/home/wind-test/Desktop/project-src/sherpa/src/llamasherpa/llamasherpa.cpp:128:11: error: use of undeclared identifier 'llama_new_context_with_model'
      ctx = llama_new_context_with_model(model, lparams);
            ^
  /usr/local/google/home/wind-test/Desktop/project-src/sherpa/src/llamasherpa/llamasherpa.cpp:131:9: error: use of undeclared identifier 'llama_free_model'
          llama_free_model(model);
          ^
  /usr/local/google/home/wind-test/Desktop/project-src/sherpa/src/llamasherpa/llamasherpa.cpp:284:17: error: unknown type name 'llama_token_data_array'; did you mean 'llama_token_data'?
                  llama_token_data_array candidates_p = { candidates.data(), candidates.size(), false };
                  ^~~~~~~~~~~~~~~~~~~~~~
                  llama_token_data
  /usr/local/google/home/wind-test/Desktop/project-src/sherpa/src/llamasherpa/../llama.cpp/llama.h:89:7: note: 'llama_token_data' declared here
      } llama_token_data;
        ^
  /usr/local/google/home/wind-test/Desktop/project-src/sherpa/src/llamasherpa/llamasherpa.cpp:284:57: error: cannot initialize a member subobject of type 'llama_token' (aka 'int') with an rvalue of type 'std::vector<llama_token_data>::value_type *' (aka 'llama_token_data *')
                  llama_token_data_array candidates_p = { candidates.data(), candidates.size(), false };
                                                          ^~~~~~~~~~~~~~~~~
  /usr/local/google/home/wind-test/Desktop/project-src/sherpa/src/llamasherpa/llamasherpa.cpp:284:76: error: non-constant-expression cannot be narrowed from type 'std::vector<llama_token_data>::size_type' (aka 'unsigned long') to 'float' in initializer list [-Wc++11-narrowing]
                  llama_token_data_array candidates_p = { candidates.data(), candidates.size(), false };
                                                                             ^~~~~~~~~~~~~~~~~
  /usr/local/google/home/wind-test/Desktop/project-src/sherpa/src/llamasherpa/llamasherpa.cpp:284:76: note: insert an explicit cast to silence this issue
                  llama_token_data_array candidates_p = { candidates.data(), candidates.size(), false };
                                                                             ^~~~~~~~~~~~~~~~~
                                                                             static_cast<float>( )
  /usr/local/google/home/wind-test/Desktop/project-src/sherpa/src/llamasherpa/llamasherpa.cpp:287:41: error: use of undeclared identifier 'llama_token_nl'
                  float nl_logit = logits[llama_token_nl()];
                                          ^
  /usr/local/google/home/wind-test/Desktop/project-src/sherpa/src/llamasherpa/llamasherpa.cpp:289:17: error: use of undeclared identifier 'llama_sample_repetition_penalty'
                  llama_sample_repetition_penalty(ctx, &candidates_p,
                  ^
  /usr/local/google/home/wind-test/Desktop/project-src/sherpa/src/llamasherpa/llamasherpa.cpp:292:17: error: use of undeclared identifier 'llama_sample_frequency_and_presence_penalties'
                  llama_sample_frequency_and_presence_penalties(ctx, &candidates_p,
                  ^
  /usr/local/google/home/wind-test/Desktop/project-src/sherpa/src/llamasherpa/llamasherpa.cpp:296:28: error: use of undeclared identifier 'llama_token_nl'
                      logits[llama_token_nl()] = nl_logit;
                             ^
  /usr/local/google/home/wind-test/Desktop/project-src/sherpa/src/llamasherpa/llamasherpa.cpp:301:26: error: use of undeclared identifier 'llama_sample_token_greedy'
                      id = llama_sample_token_greedy(ctx, &candidates_p);
                           ^
  /usr/local/google/home/wind-test/Desktop/project-src/sherpa/src/llamasherpa/llamasherpa.cpp:306:25: error: use of undeclared identifier 'llama_sample_temperature'
                          llama_sample_temperature(ctx, &candidates_p, temp);
                          ^
  /usr/local/google/home/wind-test/Desktop/project-src/sherpa/src/llamasherpa/llamasherpa.cpp:307:30: error: use of undeclared identifier 'llama_sample_token_mirostat'
                          id = llama_sample_token_mirostat(ctx, &candidates_p, mirostat_tau, mirostat_eta, mirostat_m, &mirostat_mu);
                               ^
  /usr/local/google/home/wind-test/Desktop/project-src/sherpa/src/llamasherpa/llamasherpa.cpp:310:25: error: use of undeclared identifier 'llama_sample_temperature'
                          llama_sample_temperature(ctx, &candidates_p, temp);
                          ^
  /usr/local/google/home/wind-test/Desktop/project-src/sherpa/src/llamasherpa/llamasherpa.cpp:311:30: error: use of undeclared identifier 'llama_sample_token_mirostat_v2'
                          id = llama_sample_token_mirostat_v2(ctx, &candidates_p, mirostat_tau, mirostat_eta, &mirostat_mu);
                               ^
  /usr/local/google/home/wind-test/Desktop/project-src/sherpa/src/llamasherpa/llamasherpa.cpp:314:25: error: use of undeclared identifier 'llama_sample_top_k'
                          llama_sample_top_k(ctx, &candidates_p, top_k, 1);
                          ^
  fatal error: too many errors emitted, stopping now [-ferror-limit=]
  20 errors generated.
  [3/5] Building CXX object llama.cpp/CMakeFiles/llama.dir/llama.cpp.o
  ninja: build stopped: subcommand failed.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
* Get more help at https://help.gradle.org

BUILD FAILED in 25s
Exception: Gradle task assembleDebug failed with exit code 1
Exited

**Comment by dsd** [comment]
Looks like you have the wrong llama.cpp available under src/
Did you initialize it from git submodules?

**Comment by windmaple** [comment]
Somehow my src folder was messed up. I downloaded your src zip file and it works now.
Great work, btw!

**Comment by dsd** [comment]
Also debug mode is slooow - remember to run with `--release` it will be much faster :)

**Comment by windmaple** [comment]
I think there is sth. missing for Mac

flutter: llamasherpa loaded
flutter: MessageNewLineFromIsolate : [isolate 09:02:55] llamasherpa loaded
flutter: filePath : /Volumes/Macintosh HD/Users/wind-test/Desktop/orca-mini-3b.ggmlv3.q4_1.bin
[ERROR:flutter/runtime/dart_isolate.cc(1097)] Unhandled exception:
Invalid argument(s): Failed to lookup symbol 'llamasherpa_start': dlsym(RTLD_DEFAULT, llamasherpa_start): symbol not found
#0      DynamicLibrary.lookup (dart:ffi-patch/ffi_dynamic_library_patch.dart:33:70)
#1      NativeLibrary._llamasherpa_startPtr
generated_bindings_llamasherpa.dart:41
#2      NativeLibrary._llamasherpa_startPtr (package:sherpa/generated_bindings_llamasherpa.dart)
generated_bindings_llamasherpa.dart:1
#3      NativeLibrary._llamasherpa_start
generated_bindings_llamasherpa.dart:42
#4      NativeLibrary._llamasherpa_start (package:sherpa/generated_bindings_llamasherpa.dart)
generated_bindings_llamasherpa.dart:1
#5      NativeLibrary.llamasherpa_start
generated_bindings_llamasherpa.dart:27
#6      Lib.binaryIsolate

I did not see similar issue on Linux.

**Comment by dsd** [comment]
Yeah, I don't have any macOS/iOS experience or devices. Do the official sherpa versions work there?
If you want to try you could look at the instructions [here](https://levelup.gitconnected.com/port-an-existing-c-c-app-to-flutter-with-dart-ffi-8dc401a69fd7) under "FFI on macOS and iOS".
You will need to build both llamasherpa and llama.cpp as mentioned, via Xcode/Runner.

Then as for this bit:

```
nativeApiLib = Platform.isMacOS || Platform.isIOS ? DynamicLibrary.process()
```

The equivalent of this bit is already handled, so you just have to include the C++ code in the build and then it might work.

**Comment by windmaple** [comment]
Yeah, the original version works on Mac, although it crashes if you run the model twice in a row, which is a separate issue.

**Comment by NandhaKishorM** [comment]
> Yeah, I don't have any macOS/iOS experience or devices. Do the official sherpa versions work there? If you want to try you could look at the instructions [here](https://levelup.gitconnected.com/port-an-existing-c-c-app-to-flutter-with-dart-ffi-8dc401a69fd7) under "FFI on macOS and iOS". You will need to build both llamasherpa and llama.cpp as mentioned, via Xcode/Runner.
> 
> Then as for this bit:
> 
> ```
> nativeApiLib = Platform.isMacOS || Platform.isIOS ? DynamicLibrary.process()
> ```
> 
> The equivalent of this bit is already handled, so you just have to include the C++ code in the build and then it might work.

In windows its showing error such as the dll library not Found. The file is "llamasherpa.dll"

**Comment by kmn1024** [comment]
Does any one have benchmarks (tokens/second) for running a 3B or 7B model on any low end device?
@dsd mentioned 3B running "quite well" on S10 in https://github.com/Bip-Rep/sherpa/pull/12#issue-1784005807; how well is that =)
@windmaple also seems to have found success on an unknown device in https://github.com/Bip-Rep/sherpa/pull/12#issuecomment-1619765764.
